cat/ptx-v6.0.cat

```diff
@@ -0,0 +1,72 @@
+PTX
+(* NVIDIA. Parallel Thread Execution ISA Version 6.0 *)
+
+// Base relations:
+// sr: same-scope
+// sync_barrier: synchronize by barriers with same logical barrier resource (barID)
+// sync_fence: synchronize by morally strong fence.sc
+
+// Scope Tags:
+// CTA: Cooperative thread array, is an array of threads that execute a kernel concurrently or in parallel.
+//      The CTA scope is the set of all threads executing in the same CTA as the current thread.
+// GPU: Graphics processing unit. The GPU scope is the set of all threads executing in the same cluster as the current thread.
+// SYS: System. The SYS scope is the set of all threads in the current program.
+
+(*******************)
+(*   Auxiliaries   *)
+(*******************)
+
+// Explicitly add transitivity for coherence since the co in PTX is not total
+let co = co+
+
+(* Events *)
+let strong-write = W & (RLX | REL)
+let strong-read = R & (RLX | ACQ)
+let strong-m = strong-write | strong-read
+let strong-operation = strong-m | F
+
+(* Relations *)
+let sr-fence = ([strong-operation]; sr; [F]) | ([F]; sr; [strong-operation])
+let po-fence = ([strong-operation]; po; [F]) | ([F]; po; [strong-operation])
+let sr-m = ([strong-m]; (sr & loc \ id); [strong-m])
+let po-m = ([M]; ((po | po^-1) & loc); [M])
+let morally-strong = sr-fence | po-fence | sr-m | po-m
```

Comment:
`po`-related events are always in the same scope, right?
Then `po-fence \subseteq sr-fence`, so the `po-fence` relation is pointless.

Reply:
While events in the same thread can have different scope tags, my understanding of `sr` is that it tells you which events can eventually synchronize. Events in the same thread should always be able to synchronize, thus it should be the case that `po \subseteq sr`. If this is true, then indeed the relation can be removed (I tried it locally and all tests pass).

@tonghaining can you please confirm if `po \subseteq sr`?

Reply:
The `sr` relation tells whether two events specifie the same scope (and the same scope index). I think for example in this thread P0:
`P0@cta 0,gpu 0; 
 st.release.cta x, 1 ; 
 fence.acq_rel.gpu ;`
 `W` and `F` are `po` but not `sr`. Removing this line will not include them in `morally-strong` relation. I think we don't have a test to show this in our test set.

Reply:
I would go with this definition that follows closely the English documentation
```
// The operations are related in program order, or each operation is strong and 
// specifies a scope that includes the thread executing the other operation.
let ms1 = po | ([strong-operation]; sr; [strong-operation])
// Both operations are performed via the same proxy.
let ms2 = same-proxy
// If both are memory operations, then they overlap completely
let ms3 = ((M * M) & vloc) | ((_ * _) \ (M * M))
let morally-strong = (ms1 & ms2 & ms3) \ id
```

Something similar would need to be done for `v6.0`

Reply:
Updated